component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/util/UserAgentParser.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.am.analytics.publisher.util;
+
+import org.apache.log4j.Logger;
+import ua_parser.Client;
+import ua_parser.Parser;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * User agent parser util class
+ */
+public class UserAgentParser {
+    private static final Logger log = Logger.getLogger(UserAgentParser.class);
+    private static final UserAgentParser INSTANCE = new UserAgentParser();
+    private boolean isInitialized = false;
+    private Parser uaParser;
+    private Map<String, Client> clientCache;
+
+    private UserAgentParser() {
+        try {
+            uaParser = new Parser();
+            isInitialized = true;
+            clientCache = new LinkedHashMap<String, Client>() {
+                static final int MAX = Constants.USER_AGENT_DEFAULT_CACHE_SIZE;
+                @Override
+                protected boolean removeEldestEntry(Map.Entry eldest) {
+                    return size() > MAX;
+                }
+            };
+        } catch (IOException e) {
+            log.error("Error occurred when initializing uaParser", e);
+        }
+    }
+
+    public static UserAgentParser getInstance() {
+        return INSTANCE;
+    }
+
+    public Client parseUserAgent(String userAgentHeader) {
+        if (isInitialized) {
+            Client client = clientCache.get(userAgentHeader);
+            if (client == null) {
+                client = uaParser.parse(userAgentHeader);
+                clientCache.put(userAgentHeader, client);
```

Comment:
can we keep the OS and user agent in the cache?

Reply:
There is no point is doing that as both of them are public class level variables of Client. retrieving them isnt costly at all. But to keep that in the cache we will have to create a wrapper object and put in here again which is as same as using Client